src/iterators/bfs.jl

```diff
@@ -74,36 +77,24 @@ end
 Iterator to visit vertices in a graph using breadth-first search.
 """
 function Base.iterate(t::BFSIterator, state::BFSVertexIteratorState)
-    graph, visited, queue = t.graph, state.visited, state.queue
-    while !isempty(queue)
-        if state.n_visited == nv(graph)
-            return nothing
-        end
-        # we visit the first node in the queue
-        node_start = first(queue)
-        if !visited[node_start]
-            visited[node_start] = true
-            state.n_visited += 1
-            return (node_start, state)
-        end
-        # which means we arrive here when the first node was visited.
-        neigh = outneighbors(graph, node_start)
-        if state.neighbor_idx <= length(neigh)
-            node = neigh[state.neighbor_idx]
-            # we update the idx of the neighbor we will visit,
-            # if it is already visited, we repeat
-            state.neighbor_idx += 1
-            if !visited[node]
-                push!(queue, node)
-                state.visited[node] = true
-                state.n_visited += 1
-                return (node, state)
+    # we fill nodes in this level
+    if state.node_idx == length(state.curr_level)
+        state.n_visited += length(state.curr_level)
+        state.n_visited == nv(t.graph) && return nothing
+        @inbounds for node in state.curr_level
```

Comment:
if the indices were not checked for correctness, `@inbounds` is dangerous

Reply:
it is checked at iterator creation, so I think it should be fine

Reply:
I don't think so because `state.curr_level` is a vector with variable length